app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt

```diff
@@ -0,0 +1,194 @@
+import com.saltedge.authenticator.sdk.model.response.SubmitActionResponseData
+import com.saltedge.authenticator.sdk.tools.keystore.KeyStoreManagerAbs
+import javax.inject.Inject
+
+class SubmitActionViewModel @Inject constructor(
+    private val appContext: Context,
+    private val connectionsRepository: ConnectionsRepositoryAbs,
+    private val keyStoreManager: KeyStoreManagerAbs,
+    private val apiManager: AuthenticatorApiManagerAbs
+) : ViewModel(), LifecycleObserver, ActionSubmitListener {
+
+    private var viewMode: ViewMode = ViewMode.START
+    private var actionAppLinkData: ActionAppLinkData? = null
+    private var connectionAndKey: ConnectionAndKey? = null
+
+    var onCloseEvent = MutableLiveData<ViewModelEvent<Unit>>()
+        private set
+    var onShowErrorEvent = MutableLiveData<ViewModelEvent<String>>()
+        private set
+    var onOpenLinkEvent = MutableLiveData<ViewModelEvent<String>>()
+        private set
+    var showConnectionsSelectorFragmentEvent = MutableLiveData<List<ConnectionViewModel>>()
+        private set
+    var showSubmitActionFragmentEvent = MutableLiveData<ViewModelEvent<ActionAppLinkData>>()
+        private set
+    var setResultAuthorizationIdentifier = MutableLiveData<AuthorizationIdentifier>()
+
+    val iconResId: MutableLiveData<Int> = MutableLiveData(R.drawable.ic_status_error)
+    val completeTitleResId: MutableLiveData<Int> = MutableLiveData(R.string.action_error_title)
+    val completeDescriptionResId: MutableLiveData<Int> = MutableLiveData(R.string.action_error_description)
+    val mainActionTextResId: MutableLiveData<Int> = MutableLiveData(R.string.actions_try_again)
+
+    var completeViewVisibility = MutableLiveData<Int>()
+    var actionProcessingVisibility = MutableLiveData<Int>()
+
+    override fun onActionInitFailure(error: ApiErrorData) {
+        showActionError(error.getErrorMessage(appContext))
+    }
+
+    override fun onActionInitSuccess(response: SubmitActionResponseData) {
+        val authorizationID = response.authorizationId ?: ""
+        val connectionID = response.connectionId ?: ""
+        if (response.success == true && authorizationID.isNotEmpty() && connectionID.isNotEmpty()) {
+            viewMode = ViewMode.ACTION_SUCCESS
+            onCloseEvent.postValue(ViewModelEvent(Unit))
+            setResultAuthorizationIdentifier.postValue(
+                AuthorizationIdentifier(
+                    authorizationID = authorizationID,
+                    connectionID = connectionID
+                )
+            )
+        } else {
+            viewMode = if (response.success == true) ViewMode.ACTION_SUCCESS else ViewMode.ACTION_ERROR
+            updateViewsContent()
+        }
+    }
+
+    fun setInitialData(actionAppLinkData: ActionAppLinkData) {
+        val connections = connectionsRepository.getByConnectUrl(actionAppLinkData.connectUrl)
+        this.connectionAndKey = connections.firstOrNull()?.let {
```

Comment:
why you set `this.connectionAndKey` if you have more than one suitable connections?